alvr/client_openxr/src/extra_extensions/spatial_marker_tracking.rs

```diff
@@ -0,0 +1,299 @@
+// Note: The Meta implementation is currently bugged and the buffer capacity cannot be changed once
+// the fist call of query_spatial_component_data is made.
+const MAX_MARKERS_COUNT: usize = 32;
+const DISCOVERY_TIMEOUT: Duration = Duration::from_secs(1);
+
+pub struct QRCodesSpatialContext {
+    instance: xr::Instance,
+    spatial_entity_fns: raw::SpatialEntityEXT,
+    inner: sys::SpatialContextEXT,
+    enabled_components: Vec<sys::SpatialComponentTypeEXT>,
+    entity_ids: [sys::SpatialEntityIdEXT; MAX_MARKERS_COUNT],
+    entity_states: [sys::SpatialEntityTrackingStateEXT; MAX_MARKERS_COUNT],
+    bounded_2d_arr: [sys::SpatialBounded2DDataEXT; MAX_MARKERS_COUNT],
+    marker_arr: [sys::SpatialMarkerDataEXT; MAX_MARKERS_COUNT],
+    string_buffer: [c_char; 256],
+    discovery_enabled: RelaxedAtomic,
+    discovery_timeout_deadline: Instant,
+    snapshot_future: Option<sys::FutureEXT>,
+}
+
+impl QRCodesSpatialContext {
+    pub fn new<G>(session: &xr::Session<G>, initial_discovery_state: bool) -> xr::Result<Self> {
+        let spatial_entity_fns = session
+            .instance()
+            .exts()
+            .ext_spatial_entity
+            .ok_or(sys::Result::ERROR_EXTENSION_NOT_PRESENT)?;
+        if session
+            .instance()
+            .exts()
+            .ext_spatial_marker_tracking
+            .is_none()
+        {
+            return Err(sys::Result::ERROR_EXTENSION_NOT_PRESENT);
+        }
+        alvr_common::error!("MarkerSpatialContext 1");
+
+        let enabled_components = vec![
+            sys::SpatialComponentTypeEXT::BOUNDED_2D,
+            sys::SpatialComponentTypeEXT::MARKER,
+        ];
+
+        let qr_code_capability_configuration = sys::SpatialCapabilityConfigurationQrCodeEXT {
+            ty: sys::SpatialCapabilityConfigurationQrCodeEXT::TYPE,
+            next: ptr::null(),
+            capability: CAPABILITY,
+            enabled_component_count: enabled_components.len() as u32,
+            enabled_components: enabled_components.as_ptr(),
+        };
+
+        let base_capability_configuration = ptr::from_ref(&qr_code_capability_configuration).cast();
+        let spatial_context_create_info = sys::SpatialContextCreateInfoEXT {
+            ty: sys::SpatialContextCreateInfoEXT::TYPE,
+            next: ptr::null(),
+            capability_config_count: 1,
+            capability_configs: &raw const base_capability_configuration,
+        };
+
+        let mut create_context_future: sys::FutureEXT = 0;
+        unsafe {
+            super::xr_res((spatial_entity_fns.create_spatial_context_async)(
+                session.as_handle(),
+                &spatial_context_create_info,
+                &mut create_context_future,
+            ))?;
+        }
+
+        loop {
+            if super::check_future(session.instance(), create_context_future)? {
+                break;
+            }
+            thread::sleep(std::time::Duration::from_millis(1));
```

Comment:
I think something like 5-10 ms is more sane...

Reply:
I guess. I just didn't want to dwell much on it because this blocks the while VR experience. I may want to do something about it because it's literally seconds